include/capture.h

```diff
@@ -0,0 +1,167 @@
+
+	subrange<Dims> get_subrange() const { return sr; }
+
+	const std::vector<T>& get_data() const { return data; }
+
+	std::vector<T> into_data() && { return std::move(data); }
+
+	inline const T& operator[](id<Dims> index) const { return data[detail::get_linear_index(sr.range, index)]; }
+
+	inline detail::subscript_result_t<Dims, const buffer_data> operator[](size_t index) const { return detail::subscript<Dims>(*this, index); }
+
+  private:
+	friend class capture<buffer<T, Dims>>;
+
+	subrange<Dims> sr;
+	std::vector<T> data;
+
+	explicit buffer_data(subrange<Dims> sr, std::vector<T> data) : sr{sr}, data{std::move(data)} { assert(this->data.size() == this->sr.range.size()); }
+};
+
+template <typename T, int Dims>
+class capture<buffer<T, Dims>> {
+  public:
+	using value_type = buffer_data<T, Dims>;
+
+	explicit capture(buffer<T, Dims> buf) : buffer{std::move(buf)}, sr{{}, buffer.get_range()} {}
+	explicit capture(buffer<T, Dims> buf, const subrange<Dims>& sr) : buffer{std::move(buf)}, sr{sr} {}
+
+  private:
+	friend struct detail::capture_inspector;
+
+	buffer<T, Dims> buffer;
+	subrange<Dims> sr;
+
+	void record_requirements(detail::buffer_capture_map& captures, detail::side_effect_map&) const {
+		captures.add_read_access(detail::get_buffer_id(buffer), detail::subrange_cast<3>(sr));
+	}
+
+	value_type exfiltrate_by_copy() const {
+		auto& bm = detail::runtime::get_instance().get_buffer_manager();
+		const auto access_info =
+		    bm.get_host_buffer<T, Dims>(detail::get_buffer_id(buffer), access_mode::read, detail::range_cast<3>(sr.range), detail::id_cast<3>(sr.offset));
+
+		// TODO this should be able to use host_buffer_storage::get_data
+		const auto allocation_window = buffer_allocation_window<T, Dims>{
+		    access_info.buffer.get_pointer(),
+		    buffer.get_range(),
+		    access_info.buffer.get_range(),
+		    sr.range,
+		    access_info.offset,
+		    sr.offset,
+		};
+		const auto allocation_range_3 = detail::range_cast<3>(allocation_window.get_allocation_range());
+		const auto window_range_3 = detail::range_cast<3>(allocation_window.get_window_range());
+		const auto read_offset_3 = detail::id_cast<3>(allocation_window.get_window_offset_in_allocation());
+		std::vector<T> data(allocation_window.get_window_range().size());
+		for(id<3> item{0, 0, 0}; item[0] < window_range_3[0]; ++item[0]) {
+			for(item[1] = 0; item[1] < window_range_3[1]; ++item[1]) {
+				for(item[2] = 0; item[2] < window_range_3[2]; ++item[2]) {
+					data[detail::get_linear_index(window_range_3, item)] =
+					    allocation_window.get_allocation()[detail::get_linear_index(allocation_range_3, item + read_offset_3)];
+				}
+			}
+		}
+
+		return value_type{sr, std::move(data)};
+	}
+
+	value_type exfiltrate_by_move() const { return exfiltrate_by_copy(); }
+};
+
+template <typename T, int Dims>
+capture(buffer<T, Dims>) -> capture<buffer<T, Dims>>;
+
+template <typename T, int Dims>
+capture(buffer<T, Dims>, subrange<Dims>) -> capture<buffer<T, Dims>>;
```

Comment:
Are these necessary?
(Probably "yes" in C++17 and "no" in 20?)

Reply:
Both, see https://godbolt.org/z/r6PcaTP4q

I believe there are no implicit CTAD guides for these constructors because they are not on the primary template.